### PR TITLE
Fix inconsistent results when parsing large durations and constructing durations from code

### PR DESCRIPTION
**Pull request by Envek**
Fixes next comparison:

    ActiveSupport::Duration.parse('P3Y') == 3.years # It should be true

Duration parsing made independent from any moment of time:
Fixed length in seconds is assigned to each duration part during parsing.

Methods on `Numeric` like `2.days` now use these predefined durations to avoid duplicating of duration constants through the codebase and eliminate creation of intermediate durations.

Partial backport of https://github.com/rails/rails/pull/27610 except changes in month and year duration lengths.

r? @pixeltrix

### CODE REVIEW

File: activesupport/lib/active_support/duration.rb [hunk 1]
```diff
@@ -7,7 +7,15 @@ module ActiveSupport
...
   class Duration
-    EPOCH = ::Time.utc(2000)
+    PARTS_IN_SECONDS = {
+      seconds:                       1, # Used in parse method for ease of handling part hashes with seconds
+      minutes:                      60,
+      hours:                   60 * 60,
+      days:               24 * 60 * 60,
+      weeks:       7    * 24 * 60 * 60,
+      months:     30    * 24 * 60 * 60,
+      years:    (365.25 * 24 * 60 * 60).ceil,
```
Comment (al2o3cr) [inline activesupport/lib/active_support/duration.rb]: It's slightly less clear, but reorganizing this as `1461 * 6 * 60 * 60` (absorbing a factor of 4 from 24 into the first term) will make it an integer like the others.

Comment (pixeltrix) [inline activesupport/lib/active_support/duration.rb]: I'm not sure whether breaking them out into calculations gives us anything - we're all taught in primary school the number of seconds in a minute, minutes in an hour, etc.

Comment (Envek) [inline activesupport/lib/active_support/duration.rb]: Replaced calculations with precalculated values with comments for months and years (as they will be changed in 5.1).

**Comment by pixeltrix** [comment]
Backported in 19c450d5d99275924254b2041b6ad470fdaa1f93 along with some related changes

**Pull request closed without merging**